### PR TITLE
Added support for generating Unity style tests

### PR DESCRIPTION
**Pull request by thk123**
Added a new class to generate Unity style tests. These include nice reporting of whether the test passes or fails as well as handling more sophisticated assertions. 

Most of the regression files are modified as the format of the test is different.

### CODE REVIEW

File: src/cbmc/bmc_cover.cpp [hunk 1]
```diff
@@ -305,9 +306,10 @@ bool bmc_covert::operator()()
   for(const auto & it : goal_map)
     if(it.second.satisfied) goals_covered++;
   
-  if (bmc.options.get_bool_option("gen-c-test-case"))
+  if (bmc.options.get_option("gen-c-test-case") != "")
```
Comment (thk123) [inline src/cbmc/bmc_cover.cpp]: Spacing

File: src/cbmc/bmc_cover.cpp [hunk 2]
```diff
@@ -305,9 +306,10 @@ bool bmc_covert::operator()()
...
   {
 
+    std::string generator_str = bmc.options.get_option("gen-c-test-case");
```
Comment (thk123) [inline src/cbmc/bmc_cover.cpp]: Spacing

File: src/cbmc/bmc_cover.cpp [hunk 3]
```diff
@@ -320,13 +322,34 @@ bool bmc_covert::operator()()
           +id2string(goal_map.at(goalid).source_location.get_line()));
       }
 
-      c_simple_test_case_generatort gen(get_message_handler(), bmc.options,
-        bmc.ns.get_symbol_table(), goto_functions, test, test_case_no, true);
+      std::unique_ptr<c_test_case_generatort> gen;
+      if(generator_str == "unity")
```
Comment (thk123) [inline src/cbmc/bmc_cover.cpp]: Spacing in this block x4

File: src/test-c-gen/c_test_case_generator.h [hunk 4]
```diff
@@ -48,11 +49,14 @@ class c_test_case_generatort : public messaget
 
 protected:
   virtual void add_includes(c_test_filet &test_file);
-  virtual void add_main_method(c_test_filet &test_file,
+  virtual void add_main_method(class c_test_filet &test_file,
```
Comment (thk123) [inline src/test-c-gen/c_test_case_generator.h]: Parameter on next line

File: src/test-c-gen/c_unity_test_case_generator.cpp [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+#include <test-c-gen/c_unity_test_case_generator.h>
+#include <test-c-gen/c_test_file.h>
+#include <test-c-gen/expr2cleanc.h>
+#include <string>
+#include <vector>
+#include <algorithm>
+#include <util/config.h>
+
+c_unity_test_case_generatort::c_unity_test_case_generatort(message_handlert &_message_handler,
```
Comment (thk123) [inline src/test-c-gen/c_unity_test_case_generator.cpp]: Does constructor need comment?

File: src/test-c-gen/c_unity_test_case_generator.cpp [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  const optionst &options,
+  const symbol_tablet &symbol_table,
+  const goto_functionst &goto_functions,
+  const testt &test,
+  size_t test_index,
+  bool using_test_main)
+  : c_test_case_generatort(_message_handler, options, symbol_table,
+      goto_functions, test, test_index, using_test_main)
```
Comment (thk123) [inline src/test-c-gen/c_unity_test_case_generator.cpp]: Each of these should be on a separate line.

File: src/test-c-gen/c_unity_test_case_generator.cpp [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+/*******************************************************************\
+Function: c_unity_test_case_generatort::add_main_method
+Inputs:
+ test_file - The C file to add the includes to
+ tests - The tests that need to be called
+Purpose: To create the main method that will call all the tests
+         included inside this file
+\*******************************************************************/
+void c_unity_test_case_generatort::add_main_method(c_test_filet &test_file,
+  const testt &test)
```
Comment (thk123) [inline src/test-c-gen/c_unity_test_case_generator.cpp:74]: Parameter on separate line.

File: src/test-c-gen/c_unity_test_case_generator.cpp [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+/*******************************************************************\
+Function: c_unity_test_case_generatort::add_unity_assert
+Inputs:
+ test_file - The C test file being created
+ correct_expression - The expression representing the value expected
+ return_value_var - The name of the variable (including relevant nesting)
+Purpose: Add an assertion for a unity expression
+ \*******************************************************************/
+void c_unity_test_case_generatort::add_simple_assert(c_test_filet &test_file,
```
Comment (thk123) [inline src/test-c-gen/c_unity_test_case_generator.cpp]: Parameter on separate line.

File: src/test-c-gen/c_unity_test_case_generator.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+#define CPROVER_TEST_C_GEN_C_UNITY_TEST_CASE_GENERATOR_H
+
+#include <test-c-gen/c_test_case_generator.h>
+#include <string>
+#include <vector>
+
+class c_unity_test_case_generatort : public c_test_case_generatort
+{
+public:
+  c_unity_test_case_generatort(message_handlert &_message_handler,
```
Comment (thk123) [inline src/test-c-gen/c_unity_test_case_generator.h]: Parameter on new line

File: src/test-c-gen/c_unity_test_case_generator.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    const class optionst &options,
+    const class symbol_tablet &symbol_table,
+    const class goto_functionst &goto_functions,
+    const testt &test,
+    size_t test_index,
+    bool using_test_main);
+
+protected:
+  virtual void add_includes(class c_test_filet &test_file) override;
+  virtual void add_main_method(class c_test_filet &test_file,
```
Comment (thk123) [inline src/test-c-gen/c_unity_test_case_generator.h]: Parameter on new line x4

**Review by thk123 — COMMENTED** [review]
A checklist of newly introduced whitespace changes I should fix. Can't make it a request changes on my own PR...

File: src/test-c-gen/c_unity_test_case_generator.cpp [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+\*******************************************************************/
+c_unity_test_case_generatort::c_unity_test_case_generatort(
+  message_handlert &_message_handler,
+  const optionst &options,
+  const symbol_tablet &symbol_table,
+  const goto_functionst &goto_functions,
+  const testt &test,
+  size_t test_index,
+  bool using_test_main)
+  : c_test_case_generatort(
```
Comment (peterschrammel) [inline src/test-c-gen/c_unity_test_case_generator.cpp]: put colon on the previous line; "c_test_case_generator(" is aligned with bool; its arguments are indented (as they are already)

**Review by peterschrammel — CHANGES_REQUESTED** [review]
very nice

**Pull request closed without merging**